spec/requests/check_link_spec.rb

```diff
@@ -0,0 +1,115 @@
+require "rails_helper"
+
+RSpec.describe "check link path", type: :request do
+  let(:link) { "http://www.example.com" }
+
+  def check_link_path(query_params = {})
+    path = "/check-link"
+    path + (query_params.empty? ? "?#{query_params.to_query}" : "")
+  end
+
+  def build_link_report(params)
+    {
+      "uri"       => params.fetch(:uri, anything),
+      "status"    => params.fetch(:status, anything),
+      "checked"   => params.fetch(:checked, anything),
+      "errors"    => params.fetch(:errors, []),
+      "warnings"  => params.fetch(:warnigns, []),
+    }
+  end
+
+  shared_examples "returns link report" do
+    it "returns 200" do
+      expect(response).to have_http_status(200)
+    end
+
+    it "returns a link report" do
+      json = JSON.parse(response.body)
+      expect(json).to match(link_report)
+    end
+  end
+
+  context "when no uri is requested" do
+    before do
+      get "/check-link"
+    end
+
+    it "returns 400" do
+      expect(response).to have_http_status(400)
+    end
+  end
+
+  context "when an ok checked uri is requested" do
+    let(:link_report) { build_link_report(uri: link, status: "ok") }
+
+    before do
+      # create db entries
+      get check_link_path(uri: link)
+    end
+
+    include_examples "returns link report"
+  end
+
+  context "when an unchecked uri is requested" do
+    let(:link_report) { build_link_report(uri: link, status: "pending") }
+
+    before { get check_link_path(uri: link) }
+
+    include_examples "returns link report"
+  end
+
+  context "when a checked uri, that is of status caution, is requested" do
+    let(:warnings) do
+      {
+        "risky_tld" => ["Potentially suspicious top level domain."],
```

Comment:
I don't know if we should be testing the exact contents of the human readable strings as they might want to change reasonably frequently, whereas the error type slug will never change.